kusto/internal/frames/v2/decoder.go

```diff
@@ -164,3 +161,46 @@ func (d *Decoder) decodetoMap(ctx context.Context, ch chan frames.Frame) error {
 	}
 	return nil
 }
+
+var (
+	frameType = []byte(fmt.Sprintf("%q:", frames.FieldFrameType))
+	comma     = []byte(`,`)
+	semicolon = []byte(`:`)
+)
+
+// var frameTypeRE = regexp.MustCompile(`"FrameType"\s*:\s*"([a-zA-Z]+)"`)
+
+// getFrameType looks through a raw frame to extract the type of frame. This allows us to decode the frame
+// without decoding to a map first.
+// Note: This is a fast implemenation that is benchmarked, as it is on the hot path. But it is not the
+// most robust. If we get problems, we can uncomment var frameTypeRE and the code below to do this. It takes
+// 5x as long, but in the scheme it won't matter.
+func getFrameType(message json.RawMessage) ([]byte, error) {
```

Comment:
That is mostly it.  

We do a first pass on the RawMessage, which while written to do more, really just reads the first field.  So we don't quite have to read the whole frame.

Before we had to read the entire message into a map[string]interface{}.  This was slightly faster with decoding, because it was decoded in one pass.  

But then we had create do a bunch of manual decoding of the message and type conversion from the interface{} to the concrete type.  This wasn't slow, but it does require more allocs and the code is messier.  You lose a lot of the automatic conversion you want to have.